app/models/repository.rb

```diff
@@ -29,12 +29,17 @@
 
 class Repository < ActiveRecord::Base
   include Redmine::Ciphering
+  include OpenProject::Scm::ManageableRepository
 
   belongs_to :project
   has_many :changesets, order: "#{Changeset.table_name}.committed_on DESC, #{Changeset.table_name}.id DESC"
 
   before_save :sanitize_urls
 
+  # Managed repository lifetime
+  after_save :create_managed_repository, if: Proc.new { |repo| repo.managed? }
+  after_destroy :delete_managed_repository, if: Proc.new { |repo| repo.managed? }
```

Comment:
True, only the precondition checking ahead of the delayed job is wrapped in that transaction.
Still, this provides more immediate error notification to the user than the controller running save and then never receiving any feedback due to the decoupling.

This is tightly connected to the information we can receive from delayed job. Perhaps we should avoid the delayed job completely for this task and let the user wait for the repository to be created.
(It only creates or removes one directory anyway).

We could then let a service class create a transaction and provide some kind of migration-style functionality (up, down) that will service both for the regular creating / removal flow, as well as rolling back from an error in the transaction.